feat(lua): add vim.iter

gpanders opened the pull request:
vim.iter wraps a table or iterator function into an `Iter` object with
methods such as `filtermap`, `nth`, `rev`, and `collect` which can be
chained to produce iterator pipelines that do not create new tables at
each step.

Examples:

```lua
local it1 = vim.iter({ 1, 2, 3, 4, 5 })

-- This function is evaluated "lazily", i.e. not until `collect()`, `foreach()`,
-- `next()`, or `nth()` is called, or the iterator is used in a for loop.
it1 = it1:map(function(i, v)
  if i % 2 ~= 0 then
    return i, v * 2
  end
end)

local acc = 0
for _, v in it1 do
  acc = acc + v
end
print(v) -- prints 18

-- Alternatively, use fold
-- print(it1:fold(0, function(acc, v) return acc + v end))

local it2 = vim.iter(vim.gsplit("Hello world goodbye", " ", {}))
it2 = it2:map(function(s)
  if s:match('^[a-z]') then
    return s:upper()
  end
end)
vim.print(it2:totable()) -- prints { "WORLD", "GOODBYE" }

for i, v in vim.iter({ 1, 2, 3, 4, 5 }):enumerate():rev() do
  print(string.format("%d: %d", i, v))
end
-- 1: 5
-- 2: 4
-- 3: 3
-- 4: 2
-- 5: 1
```

gpanders committed:
feat(lua): add vim.iter

vim.iter wraps a table or iterator function into an `Iter` object with
methods such as `filter`, `map`, `fold`, and `collect` which can be
chained to produce iterator pipelines that do not create new tables at
each step.

gpanders committed:
Format

gpanders committed:
Update docstrings

gpanders committed:
More docstring updates and test cases

ii14 commented:
I don't really like that you need have to call `:collect()` as the last step to process everything. If it was processing a range/array instead of iterators it could execute steps in the pipeline immediately, and it would be possible to also add `:reverse()` and `:sort()` methods.

gpanders commented:
>I don't really like that you need have to call :collect() as the last step to process everything

You need something like this to convert from the `Iter` object back into the actual table. Otherwise you wouldn't be able to chain the calls together. If `vim.iter(t):filter()` returns the filtered table, then you cannot do e.g. `vim.iter(f):filter():map()`.

gpanders committed:
Add next(), rev(), and sort()

gpanders committed:
Support standard iterator syntax

gpanders commented:
More examples:

```lua
for _, v in vim.iter({ 1, 2, 3 }):map(function(_, v) return 3 * v end):rev() do
  print(v)
end
-- 9
-- 6
-- 3
```

```lua
local it = vim.iter(string.gmatch('1,4,lol,17,blah,2,9,3', '%d+')):map(function(s) return tonumber(s) end):sort(function(a, b) return a < b)
vim.print(it.collect())
-- { 1, 2, 3, 4, 9, 17 }
```

gpanders committed:
Update test case

gpanders committed:
Use tonumber directly instead of anonymous function wrapper

gpanders committed:
Format

gpanders committed:
Remove #test tag

gpanders committed:
Remove sort(), improve reverse(), combine filter and map

gpanders committed:
Add sort option to collect. Update test cases

gpanders committed:
More test cases

gpanders committed:
Add any(), all(), last() and examples to doc comments

gpanders committed:
Add test case for enumerate()

lewis6991 commented:
I'd be interested to know some rough perf comparisons of this. If no one beats me to it I'll add some next week.

mfussenegger commented on runtime/lua/vim/Iter.lua:
```diff
@@ -0,0 +1,332 @@
+---                            in the pipeline as arguments and returns a new value. Nil values
+---                            returned from `f` are filtered from the output.
+---@return Iter
+function Iter.filter_map(self, f)
```
What's the reason to merge map and filter into one? In pretty much any language/stdlib these are separate functions.

clason replied:
Some languages have [transducers](https://juliafolds.github.io/Transducers.jl/dev/explanation/comparison_to_iterators/#comparison-to-iterators), too :P

gpanders replied:
On further thought I think `filter()` is worth having, so I brought it back.

clason replied:
> In Lua, we cannot have nil values in the middle of a table,

That's not strictly speaking correct; `{ 1, 2, nil, 4 }` is a perfectly valid table (and array but not list).

mfussenegger reviewed (commented):
Nice.

I didn't look into the details yet, from a high level perspective two questions come to mind:

- Will this eventually replace `tbl_map` / `tbl_filter`, or is this an addition?
- Should it be possible to call these functions on regular Lua iterators, directly, or does it always have to go through `vim.iter` ? E.g. compare to https://luafun.github.io/index.html


I think Justin already spent some time thinking about this, so I'd in any case wait for his feedback.

gpanders commented:
> I'd be interested to know some rough perf comparisons of this. If no one beats me to it I'll add some next week.

Initial benchmarks don't look good, at least for tables.

<details>
<summary>Script</summary>

```lua
local table_size = 100000
local test_cases = 100

local function mean(t)
  assert(#t > 0)
  local sum = 0
  for _, v in ipairs(t) do
    sum = sum + v
  end
  return sum / #t
end

local function median(t)
  local len = #t
  if len % 2 == 0 then
    return t[len/2]
  end
  return t[(len + 1)/2]
end

local function create_table()
  local t = {}
  for i = 1, table_size do
    t[#t + 1] = i
  end
  return t
end

local function test_iter()
  local t = create_table()

  local stats = {}
  for _ = 1, test_cases do
    local tic = vim.loop.hrtime()
    vim.iter(t):filter_map(function(i, v)
      if i % 2 == 0 then
        return i, v * 2
      end
    end):skip(1000):rev():collect()
    local toc = vim.loop.hrtime()
    stats[#stats + 1] = (toc - tic) / 1000000
  end
  table.sort(stats)
  return stats
end

local function test_tables()
  local t = create_table()
  local stats = {}

  for _ = 1, test_cases do
    local tic = vim.loop.hrtime()
    local mapped = vim.tbl_map(function(v) return v * 2 end, t)
    local skipped = {}
    for i = 1001, #mapped do
      skipped[#skipped + 1] = mapped[i]
    end
    local reversed = {}
    for i = 1, #skipped do
      reversed[i] = skipped[#skipped - i + 1]
    end
    local toc = vim.loop.hrtime()
    stats[#stats + 1] = (toc - tic) / 1000000
  end
  table.sort(stats)
  return stats
end

local stats_iter = test_iter()
local stats_tables = test_tables()

print(string.format("Iter: Min: %0.6f, Max: %0.6f, Median: %0.6f, Mean: %0.6f", math.min(unpack(stats_iter)), math.max(unpack(stats_iter)), median(stats_iter), mean(stats_iter)))
print(string.format("Tables: Min: %0.6f, Max: %0.6f, Median: %0.6f, Mean: %0.6f", math.min(unpack(stats_tables)), math.max(unpack(stats_tables)), median(stats_tables), mean(stats_tables)))
```

</details>

```
Iter: Min: 28.007209, Max: 33.28...edian: 32.790875, Mean: 32.689433
Tables: Min: 4.820125, Max: 5.92... Median: 4.924209, Mean: 4.985749
```

Likely because `vim.iter` iterates through the underlying table using a function, while creating new tables at each step (the `test_tables()` function in the script above) uses plain for loops and table indexes.

This suggests that it might be more efficient to convert everything to a table under the hood and mutate the underlying table at each pipeline stage. This would eliminate the "lazy evaluation" aspect though. Not sure how much that matters.

gpanders committed:
Rename Iter.fn to Iter.next

gpanders commented:
It will make the code a little more complex but I think I can special cases iterators over tables to get the best of both worlds (fast performance with tables, lazy evaluation with functional iterators).

gpanders commented:
Looking better now 😊 

```
Iter: Min: 2.716917, Max: 4.652875, Median: 2.753750, Mean: 2.853817
Tables: Min: 3.132167, Max: 3.70... Median: 3.181708, Mean: 3.228901
```

Interestingly, `enumerate()` seems to slow things down quite a bit.

Emulating enumerate with a closure:
```lua
    local i = 0
    local t = vim.iter(t):filter_map(function(v)
      i = i + 1
      if i % 2 == 0 then
        return v * 2
      end
    end):skip(1000):rev():collect()
```

is about 4 times faster than using `enumerate()`:

```lua
    local t = vim.iter(t):enumerate():filter_map(function(i, v)
      if i % 2 == 0 then
        return v * 2
      end
    end):skip(1000):rev():collect()

```

`enumerate()` is just a special case of `filter_map` that converts a table like `{4, 5, 6, 7}` to `{ {1,4}, {2, 5}, {3, 6}, {4, 7} }`. Then in subsequent pipeline stages, each table element is unpacked (note `function(i, v)` in the `enumerate` example vs `function(v)` without `enumerate`).

I think this is probably ok as-is, but we could look into optimizing that more.

gpanders committed:
Special case implementation for table iterators

gpanders commented:
> * Will this eventually replace `tbl_map` / `tbl_filter`, or is this an addition?

I think it could replace it. Unless `tbl_map` and `tbl_filter` offer a substantial benefit over `vim.iter` then it doesn't make sense to have duplicate APIs.

> * Should it be possible to call these functions on regular Lua iterators, directly, or does it always have to go through `vim.iter` ? E.g. compare to https://luafun.github.io/index.html

Regular Lua iterators are just functions, so as far as I know it's not possible to do this without first passing through `vim.iter`. But `vim.iter` does support function iterators: e.g. `vim.iter(string.gmatch("abcdefg", "%w"))`.

gpanders committed:
Implement peek, find, rfind, next_back, peek_back

gpanders committed:
Implement skip_back

gpanders committed:
Implement nth_back()

gpanders committed:
Add Iter docs

gpanders committed:
Fix lint warnings

gpanders committed:
Fix help link

gpanders committed:
Restore filter()

gpanders committed:
Add news entry

seblj commented on runtime/doc/lua.txt:
```diff
@@ -2773,4 +2808,367 @@ range({spec})                                            *vim.version.range()*
+    Return: ~
+        any
+
+Iter:foreach({self}, {f})                                     *Iter:foreach()*
```
Should it maybe be `for_each`? It seems like a lot of different languages use either camelCase `forEach` or snake_case `for_each`. (Rust, JavaScript, Java, C++, etc...). There are probably different languages that uses `foreach` as well, but IMO it looks better as `for_each` as it makes the most sense as two separate words (for each) vs (foreach)

gpanders replied:
On the other hand, Lua tends to use single words instead of underscores:

* `tonumber`
* `tostring`
* `collectgarbage`
* `setmetatable`
* `loadstring`

etc.

seblj replied:
Ah yes, that's a good argument. It also seems that there are some languages that use `foreach`. It just feels a bit wrong to me, but I don't feel _super_ strongly about it, so fine with me to leave it as well

gpanders replied:
We should be consistent either way though, so either all functions should use `singleword` style or all should use `snake_case` style. I changed everything to use `singleword` style in 39f5a075bf618af730bf6667d7d5bf85cd164951 as that is "more Lua", but if the Bikeshed Brigade comes to a consensus on `snake_case` I'll change them all back (I don't feel strongly about it either way, as long as we're consistent).

clason replied:
My vote is for noodle case, for the following reason: we use snake case specifically for `verb_noun` idioms, and none of these functions have this "complex syntax" -- they're simple terms (like `foreach`, which is indeed a builtin keyword in some languages).

gpanders replied:
I like `each` too 👍

gpanders committed:
Add vim.collect, vim.filter, vim.filter_map convenience functions

gpanders committed:
Skip table copy in Iter.collect if possible

gpanders committed:
Be consistent with naming

Omit underscores in all function names. This is more "Lua style"
(builtin Lua functions use singlewords instead of under_scores). If we
don't like this style we can convert back, but we should be consistent.

gpanders commented:
There was a suggestion to also provide some easy to use convenience functions like `vim.filter` and `vim.collect` that are small wrappers around `vim.iter`. I've provided `vim.filter`, `vim.filtermap`, and `vim.collect`.

clason commented:
About the table fast path: this only works (correctly) for list-like tables, correct? If so, (how) is this validated? Probably should be _very_ clearly documented otherwise (it's a premier footgun in Lua).

Alternatively, how much effort/performance price would it cost to support generic tables?

gpanders commented:
> About the table fast path: this only works (correctly) for list-like tables, correct? If so, (how) is this validated? Probably should be _very_ clearly documented otherwise (it's a premier footgun in Lua).

Correct. It's not really validated right now, but I will take care of that.

As far as documentation, did we ever land on established language on "list-like" vs "map-like" tables?

> Alternatively, how much effort/performance price would it cost to support generic tables?

The table specific optimizations depend on knowing the length of the table and the table being continuous; that is, it _only_ works for "list-like" tables. "Map-like" tables can easily share the function iterator interface though. I'll add that along with some test cases.

clason commented on runtime/doc/lua.txt:
```diff
@@ -2773,4 +2870,367 @@ range({spec})                                            *vim.version.range()*
+    Return: ~
+        any
+
+Iter:nextback({self})                                        *Iter:nextback()*
```
On a related note, `revnext` ("reverse next") or `endnext` might be better -- unless `nextback` is already an established term in iteratorland -- as "back" isn't used anywhere when talking about iterators here.

Alternatively, isn't that just `:rev():next()`? Do we need a separate function for this?

clason replied:
(First point applies to all the `fooback` functions  -- the docs always talk about the "end of the iterator" or the "last element".)

gpanders replied:
> On a related note, `revnext` ("reverse next") or `endnext` might be better -- unless `nextback` is already an established term in iteratorland -- as "back" isn't used anywhere when talking about iterators here.

"back" is a standard term in at least [Rust](https://doc.rust-lang.org/std/iter/trait.DoubleEndedIterator.html#tymethod.next_back) and [C++](https://en.cppreference.com/w/cpp/container/vector/back).

> Alternatively, isn't that just `:rev():next()`? Do we need a separate function for this?

No, it's not quite the same. `:rev()` will reverse the order of the table. One may want to remove a value from the end of the list without reversing the order of the list.

clason commented:
> > About the table fast path: this only works (correctly) for list-like tables, correct? If so, (how) is this validated? Probably should be _very_ clearly documented otherwise (it's a premier footgun in Lua).
> 
> Correct. It's not really validated right now, but I will take care of that.

We don't validate this anywhere else, but it might indeed be good to start -- depending on performance cost. (Especially if we end up refactoring other functions on top of this API.)

> As far as documentation, did we ever land on established language on "list-like" vs "map-like" tables?

There's `:h lua-special-tbl` and the paragraph before it (which should be expanded to introduce the terminology and get its own tag), which suggests:

* **table** is the generic Lua data structure and makes no further assumption 
* **array** is a table purely indexed by integers (which may have gaps); not relevant here
* **list** is a table purely indexed by _consecutive_ integers that start from 1
* **dict** is a table purely indexed by strings

> The table specific optimizations depend on knowing the length of the table and the table being continuous; that is, it _only_ works for "list-like" tables. "Map-like" tables can easily share the function iterator interface though (it works as-is without any extra code by using `vim.iter(pairs(t))`, but we can add a quality of life improvement to use `pairs(t)` automatically if `t` is a "map-like" table).

That makes sense; it's probably worth benchmarking each true(!) `vim.tbl_foo` function to decide whether to keep its current implementation or replace with an iterator.

gpanders committed:
Support map-like tables

gpanders commented:
"Map-like" tables are supported now too:

```lua
    it('handles map-like tables', function()
      local t = { a = 1, b = 2, c = 3 }
      local it = vim.iter(t):filtermap(function(k, v)
        if v % 2 ~= 0 then
          return k:upper(), v * 2
        end
      end)
      eq({ A = 2, C = 6 }, it:collect())
    end)
```

clason commented on src/nvim/highlight_group.c:
```diff
@@ -278,21 +278,21 @@ static const char *highlight_init_both[] = {
   // LSP semantic tokens
   "default link @lsp.type.class Structure",
-  "default link @lsp.type.comment Comment",
+  "default link @lsp.type.comment @comment",
```
👀

gpanders replied:
Oops 😅

gpanders replied:
Fixed

mfussenegger reviewed (commented):
Left a couple more comments on the interface, didn't dig into code yet. Take these as mostly thinking out loud, no strong opinions on the points raised.

gpanders commented on runtime/doc/lua.txt:
```diff
@@ -1653,6 +1670,51 @@ endswith({s}, {suffix})                                       *vim.endswith()*
     Return: ~
         (boolean) `true` if `suffix` is a suffix of `s`
 
+filter({src}, {f})                                              *vim.filter()*
```
If function first, table second is more common elsewhere (which it looks like it is, judging from your examples and a few others I found), then let's go with that.

gpanders commented on runtime/doc/lua.txt:
```diff
@@ -1653,6 +1670,51 @@ endswith({s}, {suffix})                                       *vim.endswith()*
+               values should be kept in the final table
+
+    Return: ~
+        (table)
```
I think that would make this less useful. `vim.filter(myfilter, someit())` is not much less to type than `vim.iter(someit()):filter(myfilter)`.

gpanders commented on runtime/doc/lua.txt:
```diff
@@ -2817,4 +2914,367 @@ range({spec})                                            *vim.version.range()*
+
+    Parameters: ~
+      • {opts}  ?table Optional arguments:
+                • sort (boolean|function): If true, sort the resulting table
+                  before returning. If a function is provided, that function
+                  is used as the comparator function to |table.sort()|.
```
I'm ok with removing it.

gpanders commented on runtime/doc/lua.txt:
```diff
@@ -2817,4 +2914,367 @@ range({spec})                                            *vim.version.range()*
+    Return: ~
+        Iter
+
+Iter:filtermap({self}, {f})                                 *Iter:filtermap()*
```
>but I guess this is to highlight the nil value behavior, right?

Yes, that was my reasoning for including `filter` in the name.

gpanders replied:
I've actually had a change of heart. I changed it to `map`, and added a comment to the doctoring mentioning that `nil` values are removed.

The overlap of `filter` and `filtermap` felt like it had the potential to be confusing. And like `foreach` -> `each`, using `map` sidesteps the issue of `noodlecase` vs `snake_case`.

gpanders commented on runtime/doc/lua.txt:
```diff
@@ -2817,4 +2914,367 @@ range({spec})                                            *vim.version.range()*
+    Return: ~
+        any
+
+Iter:rev({self})                                                  *Iter:rev()*
```
I think `rev()` is useful to have. Unlike sorting, there is no separate `table.reverse()` function.

gpanders commented on runtime/doc/lua.txt:
```diff
@@ -2817,4 +2914,367 @@ range({spec})                                            *vim.version.range()*
+    See also: ~
+      • Iter.find
+
+Iter:skip({self}, {n})                                           *Iter:skip()*
```
`skipback` is useful for creating a slice.

```lua
-- Remove the first 6 elements and the last 3 elements
vim.iter(t):skip(6):skipback(3)
```

This could replace `vim.list_slice`.

gpanders replied:
Also added `vim.iter(t):slice()`.

gpanders committed:
Remove collect opts, rename foreach to each, restore fold, swap function and table args

mfussenegger commented on runtime/doc/lua.txt:
```diff
@@ -2817,4 +2914,367 @@ range({spec})                                            *vim.version.range()*
+                returned from the previous stage in the pipeline as arguments
+                and returns true if the predicate matches.
+
+Iter:collect({self}, {opts})                                  *Iter:collect()*
```
> I borrowed the collect terminology from Rust, and I quite like it, but I'm not sure if that's a standard term.

It's used in java too but it takes a `Collector`, which is sort of a `Reducer`: https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/stream/Stream.html#collect(java.util.stream.Collector)

If it's restricted to return a table I'd tend to go with `:totable`, same as https://luafun.github.io/reducing.html#fun.totable, but for stdlib I think something that is more generally useful would be warranted

> It may not hurt to leave it in though. I like that you can implement collect using fold, but imo it's good to have a dedicated collect function that converts an iterator into a table (it's not immediately obvious that fold would do that).

Could also still call it collect with an opts table that takes an initial value and an accumulator. They could default to an empty table and table.insert.

gpanders replied:
Renamed to `totable` in 341e7fef0d754c6fcfcff2adfabcb0f444f9ae81. Restored `fold` in eaa15d437e47c97c82af3eed7bb8defc1d884e62.

gpanders replied:
>If it's restricted to return a table I'd tend to go with :totable, same as https://luafun.github.io/reducing.html#fun.totable, but for stdlib I think something that is more generally useful would be warranted

Because tables are so central in Lua, and because conversion to a table is slightly different implementation wise for list-like tables, map-like tables, and function iterators (for both performance and semantics reasons), I think having a dedicated `totable` makes sense. For a more generic interface, we have `fold`.

gpanders committed:
Rename collect to totable

gpanders committed:
Implement slice, improve(?) error messages

mfussenegger commented on runtime/lua/vim/iter.lua:
```diff
@@ -0,0 +1,770 @@
+
+---@class Iter
+---@field next function Return the next value in the iterator
+---@field _table ?table Underlying table data (table iterators only)
```
Just FYI in case you're not aware,  lua-ls recently added support for modifiers like `private`: `@field private _table? table`

Is the ? as prefix on the type intentional?  So far we afaik use it as property name suffix or add `|nil`

gpanders replied:
Thanks, didn’t know about the private field modifiers.

The ? is intentional to indicate optional, although in this case it’s not needed anyway, as those fields should move to the ListIter docstring.

mfussenegger replied:
> The ? is intentional to indicate optional, although in this case it’s not needed anyway, as those fields should move to the ListIter docstring.

I meant if the syntax is intentional. `_table ?table` vs `_table? table`. I think so far we use the latter but I'm not sure anymore.

lewis6991 replied:
Usually `?` appears after the name or the type, not before.

So either:
```
name? type
name type?
```

clason replied:
Or explicit `table|nil`; can't go wrong with that.

mfussenegger commented on runtime/lua/vim/iter.lua:
```diff
@@ -0,0 +1,770 @@
+      if args == nil then
+        break
+      end
+      if f(unpack(args)) then
+        return unpack(args)
+      end
```
```suggestion
      local unpacked = unpack(args)
      if f(unpacked) then
        return unpacked
      end
```

Would avoid duplicate `unpack` calls. Same for a couple of the other functions. (Or is there special arg passing semantics with Lua that I'm missing?)

gpanders replied:
The pack/unpack dance is necessary because we don’t know how many args the iterator function might return.

```
local unpacked = unpack(args)
```

will drop all but the first value.

We could avoid duplicate unpacking through a helper function maybe, though I’m not sure if that’s better from a performance perspective (I’m not sure how expensive unpack is versus another function call)

mfussenegger commented on runtime/lua/vim/iter.lua:
```diff
@@ -0,0 +1,770 @@
+---@class Iter
+---@field next function Return the next value in the iterator
+---@field _table ?table Underlying table data (table iterators only)
+---@field _head ?number Index to the front of a table iterator (table iterators only)
+---@field _tail ?number Index to the end of a table iterator (table iterators only)
```
```suggestion
---@field _head? integer Index to the front of a table iterator (table iterators only)
---@field _tail? integer Index to the end of a table iterator (table iterators only)
```

mfussenegger reviewed (approved):
Nice work. Left a couple more tiny suggestions.

I leave it up to you to decide if you want to wait for more inputs/reviews.

gpanders committed:
Update docstrings, remove unused opts arg from totable

gpanders committed:
More doc updates

gpanders committed:
Moar docs

gpanders committed:
Remove pack()s in favor of vararg closures

gpanders commented:
Updated benchmark script:

<details>
<summary><code>benchmark.lua</code></summary>

```lua
local table_size = 100000
local test_cases = 100

local function mean(t)
  assert(#t > 0)
  local sum = 0
  for _, v in ipairs(t) do
    sum = sum + v
  end
  return sum / #t
end

local function median(t)
  local len = #t
  if len % 2 == 0 then
    return t[len/2]
  end
  return t[(len + 1)/2]
end

local function benchmark(...)
  local t = {}
  for i = 1, table_size do
    t[#t + 1] = i
  end

  local last = nil

  for n = 1, select('#', ...) do
    local f = select(n, ...)
    local stats = {}
    local res
    for i = 1, test_cases do
      local tic = vim.loop.hrtime()
      res = f(t)
      local toc = vim.loop.hrtime()
      if i == 1 then
        vim.print(res)
      end
      stats[#stats + 1] = (toc - tic) / 1000000
    end
    table.sort(stats)
    print(string.format("[%d] Min: %0.6f, Max: %0.6f, Median: %0.6f, Mean: %0.6f", n, math.min(unpack(stats)), math.max(unpack(stats)), median(stats), mean(stats)))

    if last ~= nil then
      assert(#res == #last)
      for i, v in ipairs(res) do
        assert(last[i] == v)
      end
    end

    last = res
  end
end

local function test_iter(t)
  local i = 0
  return vim.iter(t):filtermap(function(v)
    i = i + 1
    if i % 2 == 0 then
      return v * 2
    end
  end):skip(1000):rev():totable()
end

local function test_tables(t)
  local a = {}
  local b = {}
  for i=1, #t do
    if i % 2 == 0 then
      a[#a + 1] = 2 * t[i]
    end
  end
  for i=1001, #a do
    b[#b+1] = a[i]
  end
  local c = {}
  for i=1, #b do
    c[#c+1] = b[#b - i + 1]
  end
  return c
end

benchmark(test_iter, test_tables)
```

</details>

Performance is pretty good, more or less equal to doing everything "manually" (creating new tables using for loops) and is faster in some cases, depending on how many pipeline stages you have.

One interesting finding is that when using a list table `t`, `vim.iter(ipairs(t))` is much faster than `vim.iter(t):enumerate()`. `enumerate()` seems to be pretty slow in general, for the reason I mentioned earlier (changing the "shape" of the values mid-pipeline has to allocate extra tables when iterating over a list). I added a note to the `enumerate()` docstring mentioning this.

gpanders committed:
Rename filtermap to map

gpanders committed:
Add varargs to convenience functions

gpanders committed:
Update docs

gpanders commented on runtime/doc/lua.txt:
```diff
@@ -1843,6 +1820,29 @@ list_slice({list}, {start}, {finish})                       *vim.list_slice()*
+    Parameters: ~
+      • {f}    function(...):?any Map function. Accepts the current iterator
+               or table values as arguments and returns one or more new
+               values. Nil values are removed from the final table. the final
+               table
```
```suggestion
               values. Nil values are removed from the final table.
```

gpanders commented on runtime/lua/vim/shared.lua:
```diff
@@ -884,4 +884,112 @@ function vim.defaulttable(create)
+---
+---@param f function(...):?any Map function. Accepts the current iterator or table values as
+---                            arguments and returns one or more new values. Nil values are removed
+---                            from the final table. the final table
```
```suggestion
---                            from the final table.
```

gpanders committed:
Update runtime/lua/vim/shared.lua

gpanders committed:
Update runtime/doc/lua.txt

gpanders committed:
Clarify that rev and peek are only supported for list tables

gpanders committed:
Fix errors in docstring examples

sandersantema commented:
I hope it's fine to leave some quick feedback here for later iterations/ improvements. The naming seems inconsistent in the sense that there's both `rfind` and various `<...>back`, personally I prefer `r<...>` which is more common (according to my anecdotal evidence) and less characters to type :). 

I think I've furthermore found one limiting factor in the api, or at least for me it is.

I've got the following code which I wanted to port:
```lua
for _, path in ipairs(rtpaths) do
  -- Look on runtime path for 'lua/*.lua' files
  local path1 = table.concat({ path, ext, modfile }, sep)
  if vim.fn.filereadable(path1) == 1 then
    return path1
  end

  -- Look on runtime path for 'lua/*/init.lua' files
  local path2 = table.concat({ path, ext, fname, initfile }, sep)
  if vim.fn.filereadable(path2) == 1 then
    return path2
  end
end
```

My first thought was to `:map()` the `rtpaths` to the `path1` and `path2` above and then to use `:find(function(path) vim.fn.readable(path) == 1 end)` like so:
```lua
vim
  .iter(rtpaths)
  :map(
    function(path)
      return table.concat({ path, ext, modfile }, sep),
        table.concat({ path, ext, fname, initfile }, sep)
    end
  )
  :find(function(path) return vim.fn.filereadable(path) == 1 end)
```

Unfortunately this isn't possible since the `map` has returned a table, i.e. the shape of the iterator is `{{p1, p2}, ...}` while we would like `{p1, p2, ...}`. In this case some `flatten` functionality would be very nice I think so you can do the `filter` without having to `totable` the iterator to flatten it. Other than that what a lovely addition to the API, now I can program in neovim like I do almost anywhere else and the terse functional style fits configuring neovim well I think, thanks!

gpanders commented:
>The naming seems inconsistent in the sense that there's both rfind and various <...>back, personally I prefer r<...> which is more common (according to my anecdotal evidence) and less characters to type :).

The names were mostly borrowed from [Rust](https://doc.rust-lang.org/std/iter/trait.Iterator.html) (which shares some terminology with C++ as well, e.g. `rfind` and `back`). That decision was arbitrary, I chose that simply because it's what I'm familiar with. We can bikeshed over naming into eternity, so unless a clear (and expedient) consensus emerges, I will bias towards leaving the names as-is.

>Unfortunately this isn't possible since the map has returned a table, i.e. the shape of the iterator is {{p1, p2}, ...} while we would like {p1, p2, ...}. In this case some flatten functionality would be very nice I think so you can do the filter without having to totable the iterator to flatten it. 

The API is not complete. A `flatten` implementation would indeed be nice to have (another notable missing feature is `zip()`). This PR gets the concept out the door and into users' hands. We can and will continue to refine and extend the API.

justinmk commented on test/functional/lua/vim_spec.lua:
```diff
@@ -3029,6 +3029,360 @@ describe('lua stdlib', function()
     eq(NIL, if_nil(a))
   end)
+
+  describe('vim.iter', function()
```
probably worth a new `vim/iter_spec.lua` test file :)

justinmk commented:
Great to see this. I did spend a fair bit of time evaluating [luafun](https://github.com/luafun/luafun) for this purpose. 

> * Will this eventually replace `tbl_map` / `tbl_filter`, or is this an addition?

+1 let's plan to deprecate the tbl_xx stuff eventually.

> * Should it be possible to call these functions on regular Lua iterators, directly, or does it always have to go through `vim.iter` ? E.g. compare to https://luafun.github.io/index.html

💯  I hope we can make this work. Having to wrap everything in iter() isn't ideal, though having the _option_ of a Iter object is fine.

justinmk commented:
> Unless you namespace them separately under `vim.iter.`, but now your code looks much more verbose:

That seems perfectly fine. After all, this is just an option. The point is that `vim.iter.xx()` should all work with "iterables" of any kind: 

1. the traditional Lua iter pattern
2. tables
3. `vim.iter()` wrapper

justinmk commented:
I love pretty much all the choices in this PR. Well done!

justinmk commented on runtime/lua/vim/shared.lua:
```diff
@@ -884,4 +884,109 @@ function vim.defaulttable(create)
+---
+---@param src table|function Table or iterator.
+---@return Iter @|lua-iter|
+function vim.iter(src, ...)
```
I think these top-level functions should live on `vim.iter`. Because other than the `vim.tbl_xx` things (which will be deprecated), we don't have much precedent there. Putting these on `vim.inter` helps discovery and anyone worried about verbosity can do `local it = vim.iter`.

justinmk commented:
For `Iter:find()`, `Iter:filter()`, (do we need `Iter:reduce()` ?), should we also pass the "previous" element to the callback function? This allows the callback to compare / reason about observed states.

Created an issue: https://github.com/neovim/neovim/issues/23251

gpanders commented on runtime/doc/lua.txt:
```diff
@@ -1740,6 +1818,26 @@ list_slice({list}, {start}, {finish})                       *vim.list_slice()*
     Return: ~
         (list) Copy of table sliced from start to finish (inclusive)
 
+map({f}, {src}, {...})                                             *vim.map()*
```
> @gpanders any objection to changing this signature so that the function is the last param? That is generally the pattern we should strive for, because the function is usually the biggest part of a call (lexically).
> 
> ```
> map({src}, {f})
> ```
> 
> Though if the variadic `...` arg is needed then this can't change. Is the variadic arg part of `{src}` ?

@mfussenegger and I [discussed this as well](https://github.com/neovim/neovim/pull/23029#discussion_r1167067295). I also prefer trailing function for stylistic reasons, but as you pointed out we need the variadic params in order for this to work properly (a Lua function iterator uses multiple values to maintain its state. For example, the `pairs()` function actually returns 3 values, `:h luaref-in`).

justinmk commented on runtime/lua/vim/iter.lua:
```diff
@@ -0,0 +1,836 @@
+  error('peekback() requires a list-like table')
+end
+
+function ListIter.peekback(self)
```
Looks like the difference of this compare to `last()` is that this doesn't drain the iterator? This name seems unintuitive. Would `peeklast()` make sense?

gpanders replied:
See my [other comment](https://github.com/neovim/neovim/pull/23029#discussion_r1403454682).

gpanders commented on runtime/lua/vim/iter.lua:
```diff
@@ -0,0 +1,836 @@
+  if self._head ~= self._tail then
+    local inc = self._head < self._tail and 1 or -1
+    self._tail = self._tail - inc
+    return self._table[self._tail]
```
I mentioned in [an earlier comment](https://github.com/neovim/neovim/pull/23029#issuecomment-1513244389) that for the most part I borrowed naming/terminology from Rust's Iterator interface, which also uses `nextback` (and `skipback` etc.) That choice was mostly arbitrary since it's what I am most familiar with.

`pop` makes sense too, but imo is best when also paired with `push`. Iterators don't have `push`, but they do have `next`, so `nextback` provides symmetry.
